assets/app/features/upgrade/steps.factory.spec.js

```diff
@@ -118,5 +406,87 @@ describe('Stepes Factory', function () {
+                expect(upgradeStepsFactory.stepByState('--dummy state--')).not.toBeDefined();
+            });
+        });
+
+        describe('stepByID function', function () {
+            _.forEach(mockedInitialSteps, function (mockedStep) {
+                it('should return step with ID=' + mockedStep.id + ' when called', function () {
+                    expect(upgradeStepsFactory.stepByID(mockedStep.id)).toEqual(mockedStep);
+                });
+            });
+
+            it('should return undefined when called with nonexistent ID', function () {
+                expect(upgradeStepsFactory.stepByID(-1)).not.toBeDefined();
+                expect(upgradeStepsFactory.stepByID(9999)).not.toBeDefined();
+            });
+        });
+
+        describe('lastStateForRestore function', function () {
+            describe('when there is no stored last state', function () {
+                beforeEach(function () {
+                    localStorage.removeItem(UPGRADE_LAST_STATE_KEY);
+                });
+                it('should return "upgrade-landing" before any step', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(initialStatusData))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade-landing" after prechecks', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterPrechecks))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade-landing" during prepare', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataDuringPrepare))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade.backup" after prepare', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterPrepare))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.backup" during backup', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataDuringBackup))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.backup" after backup', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterBackup))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.administration-repository-checks" after admin repo checks', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterAdminRepoChecks))
+                        .toEqual('upgrade.administration-repository-checks');
+                });
+            });
+            describe('when stored state is invalid in given case', function () {
```

Comment:
There's nothing about step here. `upgrade-landing` (stored in `localStorage`) is invalid in this test as it doesn't match the status data passed to `lastStateForRestore()`. In such cases it will be ignored and data from function argument will be used.